File: core/src/main/java/com/starshootercity/abilities/PhantomizeOverlay.java
package com.starshootercity.abilities;

import com.destroystokyo.paper.event.player.PlayerPostRespawnEvent;
import net.kyori.adventure.key.Key;
import org.bukkit.Bukkit;
import org.bukkit.WorldBorder;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.jetbrains.annotations.NotNull;

public class PhantomizeOverlay implements DependantAbility, Listener {
    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:phantomize_overlay");
    }

    @Override
    public @NotNull Key getDependencyKey() {
        return Key.key("origins:phantomize");
    }

    @EventHandler
    public void onPhantomizeToggle(Phantomize.PhantomizeToggleEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    @EventHandler
    public void onPlayerPostRespawn(PlayerPostRespawnEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        updatePhantomizeOverlay(event.getPlayer());
    }

    private void updatePhantomizeOverlay(Player player) {
        if (getDependency().isEnabled(player)) {
            WorldBorder border = Bukkit.createWorldBorder();
            border.setCenter(player.getWorld().getWorldBorder().getCenter());
            border.setSize(player.getWorld().getWorldBorder().getSize());
            border.setWarningDistance(player.getWorld().getWorldBorder().getWarningDistance()*2);
        } else player.setWorldBorder(null);
    }
}

File: core/src/main/java/com/starshootercity/abilities/PhantomizeOverlay.kt
package com.starshootercity.abilities

import com.destroystokyo.paper.event.player.PlayerPostRespawnEvent
import com.starshootercity.OriginsReborn.Companion.NMSInvoker
import com.starshootercity.abilities.Phantomize.PhantomizeToggleEvent
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent

class PhantomizeOverlay : DependantAbility, Listener {
    override fun getKey(): Key {
        return Key.key("origins:phantomize_overlay")
    }

    override fun getDependencyKey(): Key {
        return Key.key("origins:phantomize")
    }

    @EventHandler
    fun onPhantomizeToggle(event: PhantomizeToggleEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    @EventHandler
    fun onPlayerPostRespawn(event: PlayerPostRespawnEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        updatePhantomizeOverlay(event.getPlayer())
    }

    private fun updatePhantomizeOverlay(player: Player?) {
        NMSInvoker.setWorldBorderOverlay(player, dependency.isEnabled(player))
    }
}
